src/pyggdrasil/tree_inference/_tree.py

```diff
@@ -0,0 +1,273 @@
+            root = tree_inf.adjacency_to_root_dfs(
+                adj_matrix=np.array(reorder_tree.tree_topology),
+                labels=np.array(reorder_tree.labels),
+            )
+        else:
+            root = tree_inf.adjacency_to_root_dfs(
+                adj_matrix=np.array(self.tree_topology), labels=np.array(self.labels)
+            )
+        return root
+
+    def print_topo(self):
+        """Prints the tree in a human-readable format."""
+        return self.to_TreeNode().print_topo()
+
+
+def _resort_root_to_end(tree: Tree, root: int) -> Tree:
+    """Resorts tree so that root is at the end of the adjacency matrix.
+
+    Args:
+        root: int
+            root label of the tree
+        tree: Tree
+            tree to resort
+
+    Returns:
+        tree: Tree
+    """
+    # get root index
+    root_idx = int(jnp.where(tree.labels == root)[0])
+    # get all nodes which are not root
+    non_root_idx = jnp.where(tree.labels != root)[0]
+    # get new reduced adjacency matrix
+    reduced_adj = tree.tree_topology[non_root_idx, :][:, non_root_idx]
+    # resort row of root, so that it is at the end - excluding root
+    root_row = tree.tree_topology[root_idx, non_root_idx]
+    # resort column of root, so that it is at the end
+    root_col = jnp.append(
+        tree.tree_topology[non_root_idx, root_idx],
+        tree.tree_topology[root_idx, root_idx],
+    )
+    # get new adjacency matrix
+    # TODO: use matrix assignment not append
+    new_adj = jnp.append(reduced_adj, jnp.array([root_row]), axis=0)
+    new_adj = jnp.append(new_adj, jnp.swapaxes(jnp.array([root_col]), 0, 1), axis=1)
+    # get new labels
+    new_labels = jnp.append(tree.labels[non_root_idx], tree.labels[root_idx])
+    # get new tree
+    resorted_tree = Tree(new_adj, new_labels)
+
+    return resorted_tree
+
+
+def _get_descendants(
+    adj_matrix: Array, labels: Array, parent: int, includeParent: bool = False
+) -> Array:
+    """
+    Returns a list of labels representing the descendants of node parent.
+    Used boolean matrix exponentiation to find descendants.
+
+    Complexity:
+        Naive: O(n^3 * (n-1)) where n is the number of nodes in the tree including root.
+        TODO: - Consider implementing 'Exponentiation by Squaring Algorithm'
+                    for  O(n^3 * log(m)
+              - fix conditional exponentiation for exponent < n-1
+    Args:
+    - adj_matrix: a JAX array of shape (n, n) representing the adjacency matrix
+    - labels: a JAX array of shape (n,) representing the labels of the nodes
+    - parent: an integer representing
+        the label of the node whose descendants we want to find
+
+    Returns:
+    - a JAX array of integers representing the labels of the descendants of node parent
+      in order of nodes in the adjacency matrix, i.e. the order of the labels
+      if includeParent is True, the parent is included in the list of descendants
+    """
```

Comment:
`_mcmc.py` currently used  this`get_descendants()` - a second implementation using Floyd-Warshall exists in `_simulate.py`